### ISSUE TITLE
[Bug]: torch direct ml: the procedure could not be found.

### ISSUE THREAD

**Issue by Swagarot** [lshqqytiger/stable-diffusion-webui-directml#97]
### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits

### What happened?

![Screenshot_20230502-173358_TeamViewer~2](https://user-images.githubusercontent.com/57872907/235698450-3c058796-a23e-45db-8413-f778231ca538.png)

### Steps to reproduce the problem

I simply ran webui-user.bat

### What should have happened?

It should've not given me an error and open a webui instance 

### Commit where the problem happens

271edfa8d6b66e0d0b3f0b2964f35598df69d2a5

### What platforms do you use to access the UI ?

Windows

### What browsers do you use to access the UI ?

Mozilla Firefox, Google Chrome

### Command Line Arguments

```Shell
No
```

### List of extensions

No

### Console logs

```Shell
In the image I provided
```

**Comment by FrakerKill** [comment]
Me too but I try to reinstall:

![image](https://user-images.githubusercontent.com/124603471/235704618-5d1bbf6a-5d38-4abe-ba63-237c7896ffa6.png)

**Comment by Reductito** [comment]
Exact same issue here

**Comment by triggerftu** [comment]
seems the issue is torch_directml. have anyone tried reloading the install for the requirements? also i know the main developer had mentioned reinstalling torch_directml in another issue but seems not possible.

**Comment by gringocharlie** [comment]
same issue here ...

**Comment by FrakerKill** [comment]
> seems the issue is torch_directml. have anyone tried reloading the install for the requirements? also i know the main developer had mentioned reinstalling torch_directml in another issue but seems not possible.

I reinstall all and nothing.

**Comment by triggerftu** [comment]
> 

same thing. trying different things to see if i can get anything to load.

**Comment by Zaakh** [comment]
Same issue here. Had to activate my venv `.\venv\Scripts\activate` and manualy downgrade torch-directml with `pip install  torch-directml==0.1.13.1.dev230413`

**Comment by FrakerKill** [comment]
> Same issue here. Had to activate my venv `.\venv\Scripts\activate` and manualy downgrade torch-directml with `pip install torch-directml==0.1.13.1.dev230413` to fix it.

Works for me. Could be hardcoded the version of torch-directml in anywhere?

**Comment by triggerftu** [comment]
so my solution so far. loading the command `pip install -r requirements.txt` then loading the webui-user.bat got it up and running on my end.

**Comment by Zaakh** [comment]
> I tried downgrading didn't seem to work, I'll try the version mentioned above, what do you mean by activating the venv?

I use the usual venv installation, which means I have a local folder named `venv` where all python packages are installed. To follow my suggestion, open a powershell, navigate to your installation folder and execute the two commands I described.

**Comment by JJGall** [comment]
> Tried pip install torch-directml==0.1.13.1.dev230413 at the installation folder still doesn't work, if it helps when I did the command I saw a lot of requirement already satisfied

go to venv\scripts open powershell\terminal and type activate and it will show (venv) before directory then try again, or type pip show torch-directml and post what version it shows you

**Comment by Swagarot** [comment]
> > Tried pip install torch-directml==0.1.13.1.dev230413 at the installation folder still doesn't work, if it helps when I did the command I saw a lot of requirement already satisfied
> 
> go to venv\scripts open powershell\terminal and type activate and it will show (venv) before directory then try again, or type pip show torch-directml and post what version it shows you

Okay this worked, thanks everyone

**Comment by MihailCelProst** [comment]
> > Tried pip install torch-directml==0.1.13.1.dev230413 at the installation folder still doesn't work, if it helps when I did the command I saw a lot of requirement already satisfied
> 
> go to venv\scripts open powershell\terminal and type activate and it will show (venv) before directory then try again, or type pip show torch-directml and post what version it shows you

Sorry, I am quite new to all this. What do you exactly mean by powershell/terminal, I don't quite understand how I should "activate" scripts through them. And is it correct that after I "activated" scripts I need to run the webui-user.bat with the code:
`pip install torch-directml==0.1.13.1.dev230413
@echo off

set PYTHON=
set GIT=
set VENV_DIR=
set COMMANDLINE_ARGS=

call webui.bat`

Thanks in advance.

**Comment by JJGall** [comment]
> > > Tried pip install torch-directml==0.1.13.1.dev230413 at the installation folder still doesn't work, if it helps when I did the command I saw a lot of requirement already satisfied
> > 
> > 
> > go to venv\scripts open powershell\terminal and type activate and it will show (venv) before directory then try again, or type pip show torch-directml and post what version it shows you
> 
> Sorry, I am quite new to all this. What do you exactly mean by "powershell/terminal", I don't quite understand how I should "activate" scripts through them. And is it correct that after I "activated" scripts I need to run the webui-user.bat with the code:
> 
> ```
> pip install torch-directml==0.1.13.1.dev230413
> @echo off
> 
> set PYTHON=
> set GIT=
> set VENV_DIR=
> set COMMANDLINE_ARGS=
> 
> call webui.bat
> ```
> 
> Thanks in advance.

Don't put that in the webui-user.bat, if that's what you did. In the `stable-diffusion-webui-directml` there is a folder called `venv` and inside that, one called `scripts`, open the scripts folder, right click or shift + right click in the folder in some empty space, and it should have open in terminal or open powershell here, it depends on what Windows version you have, after that type `.\activate` and hit enter. It should show **(venv)** before the folder name now, you activated the `venv` don't worry about what that means. Now you can try `pip install torch-directml==0.1.13.1.dev230413` 
![venvscreen](https://user-images.githubusercontent.com/25047164/235808516-6ce42d63-2333-4450-8dc0-5d782821c764.jpg)

Hope my way of explaining it wasn't too complicated.

**Comment by revons1984** [comment]
> so my solution so far. loading the command `pip install -r requirements.txt` then loading the webui-user.bat got it up and running on my end.

Did this solution let you run it with the torch 2.0 or was this to allow you to downgrade back?

**Comment by MihailCelProst** [comment]
>> _Don't put that in the webui-user.bat . . . when the code owner straightens it out._

Thank you very much, I finally got the webui to work!

> Don't put that in the webui-user.bat, if that's what you did.

I hadn't yet, I left it in the code for demonstrational purposes.

> shift + right click in the folder in some empty space, and it should have open in terminal or open powershell here

Shift + rmb worked for me. I never knew about being able to open powershell in a specific folder, before your instructions I was trying to run the code through the powershell I opened via the windows search bar xD

However when I was attempting to run the code I had to change the execution policy, apparently I didn't have 'running scripts' enabled for my personal user account. I enabled it by following the [top answer](https://stackoverflow.com/a/67553273) on a stackoverflow post.

> you activated the venv don't worry about what that means

I'll take you word for it then lol.

> Hope my way of explaining it wasn't too complicated.

Your instructions were very clear, even for someone with only basic computer knowledge like me! The only thing that I was a little confused about was where I had to put `pip install torch-directml==0.1.13.1.dev230413`, but I quickly figured out I had to put it directly after the code (from your picture) with `(venv)` in front of it .

> Just keep in mind this might change and you may need to update again when the code owner straightens it out.

I will do, but nevertheless thanks a ton for helping me and teaching me new stuff @JJGall  !

**Comment by triggerftu** [comment]
> > so my solution so far. loading the command `pip install -r requirements.txt` then loading the webui-user.bat got it up and running on my end.
> 
> Did this solution let you run it with the torch 2.0 or was this to allow you to downgrade back?

it downgraded the torch to ver 1.13 as thats whats in the requirements. basicly the command tells python to install all that is listed and the version if it's stated in the txt file. 

also for the time torch 1.13 has to be used for AMD users as torch 2.0.0 has issues with the hardware.

**Comment by FrakerKill** [comment]
> Should I close this saying the current solution is to do the venv thing?

Now, after today's commit it's working fine

**Comment by Swagarot** [comment]
> > Should I close this saying the current solution is to do the venv thing?
> 
> Now, after today's commit it's working fine

So it has been fixed?

**Comment by Swagarot** [comment]
> Still broke guys?
> 
> `AttributeError: module 'torch' has no attribute 'ops'`

Seems to be a different error

**Comment by songib** [comment]
> > Still broke guys?
> > `AttributeError: module 'torch' has no attribute 'ops'`
> 
> Seems to be a different error

I already do the pip requirement stuff and at the end of the tunnels, I got this instead.
and some of the requirements got a conflict with sd-dynamic-prompt in the process, idk how to resolve that either. 
seems like a good day so far. xd

**Comment by Swagarot** [comment]
> > > Still broke guys?
> > > `AttributeError: module 'torch' has no attribute 'ops'`
> > 
> > 
> > Seems to be a different error
> 
> I already do the pip requirement stuff and at the end of the tunnels, I got this instead. and some of the requirements got a conflict with sd-dynamic-prompt in the process, idk how to resolve that either. seems like a good day so far. xd

I think you should open a new issue since ur problem is with just torch not torch-directml, as for how to fix maybe reinstall just torch.

**Comment by songib** [comment]
@Swagarot  I'm on directml issue, so I use the same thing as you guys, AMD. 
since most likely they try a new torch implementation.

**Comment by Swagarot** [comment]
> @Swagarot I'm on directml issue, so I use the same thing as you guys, AMD. since most likely they try a new torch implementation.

I guess we'll, I'll leave this open for now

**Comment by songib** [comment]
@Swagarot thanks
got something like this if you can help with this

Traceback (most recent call last):
  File ".stable-diffusion-webui-directml\launch.py", line 354, in <module>
    prepare_environment()
  File ".stable-diffusion-webui-directml\launch.py", line 227, in prepare_environment
    import torch_directml
  File ".stable-diffusion-webui-directml\venv\lib\site-packages\torch_directml\__init__.py", line 13, in <module>
    torch.ops.load_library(directml_dll)
AttributeError: module 'torch' has no attribute 'ops'

so yeah, should I just downgrade to this instead "torch-directml==0.1.13.1.dev230413" ?

**Comment by Swagarot** [comment]
> @Swagarot thanks got something like this if you can help with this
> 
> Traceback (most recent call last): File ".stable-diffusion-webui-directml\launch.py", line 354, in prepare_environment() File ".stable-diffusion-webui-directml\launch.py", line 227, in prepare_environment import torch_directml File ".stable-diffusion-webui-directml\venv\lib\site-packages\torch_directml__init__.py", line 13, in torch.ops.load_library(directml_dll) AttributeError: module 'torch' has no attribute 'ops'
> 
> so yeah, should I just downgrade to this instead "torch-directml==0.1.13.1.dev230413" ?

Worth a shot, follow the comments above with the venv and see if it works

**Comment by songib** [comment]
> 

ok, I tried, still the same stuff. 

Already up to date.
venv ".\stable-diffusion-webui-directml\venv\Scripts\Python.exe"
Installing DirectML
Traceback (most recent call last):
  File ".\stable-diffusion-webui-directml\launch.py", line 354, in <module>
    prepare_environment()
  File ".\stable-diffusion-webui-directml\launch.py", line 227, in prepare_environment
    import torch_directml
  File ".\stable-diffusion-webui-directml\venv\lib\site-packages\torch_directml\__init__.py", line 13, in <module>
    torch.ops.load_library(directml_dll)
AttributeError: module 'torch' has no attribute 'ops'
Press any key to continue . . .

did yours get fix?

**Comment by Swagarot** [comment]
> > 
> 
> ok, I tried, still the same stuff.
> 
> Already up to date. venv ".\stable-diffusion-webui-directml\venv\Scripts\Python.exe" Installing DirectML Traceback (most recent call last): File ".\stable-diffusion-webui-directml\launch.py", line 354, in prepare_environment() File ".\stable-diffusion-webui-directml\launch.py", line 227, in prepare_environment import torch_directml File ".\stable-diffusion-webui-directml\venv\lib\site-packages\torch_directml__init__.py", line 13, in torch.ops.load_library(directml_dll) AttributeError: module 'torch' has no attribute 'ops' Press any key to continue . . .
> 
> did yours get fix?

Mine got fixed just by downgrading torch-directml

**Comment by songib** [comment]
@Swagarot  you can close it [I Use this method and it works](https://github.com/lshqqytiger/stable-diffusion-webui-directml/issues/101#issuecomment-1532925139)

**Comment by killpotts** [comment]
> pip install torch-directml==0.1.13.1.dev230413

Tried this and still getting the same error.

**Comment by Swagarot** [comment]
> > pip install torch-directml==0.1.13.1.dev230413
> 
> Tried this and still getting the same error.
Try to follow these steps and let me know if it works,
Also if it doesn't please post a picture of the exact error you get.



> Don't put that in the webui-user.bat, if that's what you did. In the `stable-diffusion-webui-directml` there is a folder called `venv` and inside that, one called `scripts`, open the scripts folder, right click or shift + right click in the folder in some empty space, and it should have open in terminal or open powershell here, it depends on what Windows version you have, after that type `.\activate` and hit enter. It should show **(venv)** before the folder name now, you activated the `venv` don't worry about what that means. Now you can try `pip install torch-directml==0.1.13.1.dev230413` ![venvscreen](https://user-images.githubusercontent.com/25047164/235808516-6ce42d63-2333-4450-8dc0-5d782821c764.jpg)
> 
> Hope my way of explaining it wasn't too complicated. Just keep in mind this might change and you may need to update again when the code owner straightens it out.

**Comment by Vexiis** [comment]
I also am getting this error, I reverted to pytorch 1.13 and torch-directml 1.13.1.dev230413 it however did not resolve the issue.
I tried also reverting to the previous snapshot before the update to pytorch 2.0 but that didn't do it either.

**Comment by gringocharlie** [comment]
I got mine to work with a combination of two solutions above.

I installed the requirements : pip install -r requirements.txt

Then I navigated to my venv folder and executed activate

**Comment by steppol99** [comment]
Nevermind! i deleted the **venv** folder and opened again the **webui-user.bat**, it reinstalled all the **venv** folder again and now it works!

**Comment by ExRed** [comment]
I had the same issue and deleting the venv-folder did fix it for me, too.

**Comment by imfine3keyou** [comment]
Just delete the venv folder and start webui again, I've got it running successfully

**Comment by iSpockMan** [comment]
> > > > Tried pip install torch-directml==0.1.13.1.dev230413 at the installation folder still doesn't work, if it helps when I did the command I saw a lot of requirement already satisfied
> > > 
> > > 
> > > go to venv\scripts open powershell\terminal and type activate and it will show (venv) before directory then try again, or type pip show torch-directml and post what version it shows you
> > 
> > 
> > Sorry, I am quite new to all this. What do you exactly mean by "powershell/terminal", I don't quite understand how I should "activate" scripts through them. And is it correct that after I "activated" scripts I need to run the webui-user.bat with the code:
> > ```
> > pip install torch-directml==0.1.13.1.dev230413
> > @echo off
> > 
> > set PYTHON=
> > set GIT=
> > set VENV_DIR=
> > set COMMANDLINE_ARGS=
> > 
> > call webui.bat
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Thanks in advance.
> 
> Don't put that in the webui-user.bat, if that's what you did. In the `stable-diffusion-webui-directml` there is a folder called `venv` and inside that, one called `scripts`, open the scripts folder, right click or shift + right click in the folder in some empty space, and it should have open in terminal or open powershell here, it depends on what Windows version you have, after that type `.\activate` and hit enter. It should show **(venv)** before the folder name now, you activated the `venv` don't worry about what that means. Now you can try `pip install torch-directml==0.1.13.1.dev230413` ![venvscreen](https://user-images.githubusercontent.com/25047164/235808516-6ce42d63-2333-4450-8dc0-5d782821c764.jpg)
> 
> Hope my way of explaining it wasn't too complicated. Just keep in mind this might change and you may need to update again when the code owner straightens it out.

I am not entirely sure what I am doing wrong. When I launch the `web-user.bat`, I get am error about torch not being able to install correctly. I received help on Discord and I was told to type `pip install torch-directml==0.1.13.1.dev230413` in cmd inside the stable-diffusion-webui-directml directory, but that did not work. I followed the steps in this thread. I went to my `venv` folder and activated it using powershell. After activating it, I pasted `pip install torch-directml==0.1.13.1.dev230413` inside powershell and saw the download commencing. When that finished, I launched the `web-user.bat` and saw that it was downloading more this time and skipped my torch installation. I assumed it was going to work this time until I got another error. I was reading the most recent comments in this thread and people said deleting the venv folder worked for them, but it put me back to where I started which was cmd not being able to download torch. I have included screenshots of my errors below. Hopefully someone can help. 

(What happens when I launch the .bat without the venv folder)
https://imgur.com/a/u2dA58Z

(What happens when I launch the .bat after venv has been activated) 
https://imgur.com/a/7XHLSQa

**Comment by iSpockMan** [comment]
> > > > Tried pip install torch-directml==0.1.13.1.dev230413 at the installation folder still doesn't work, if it helps when I did the command I saw a lot of requirement already satisfied
> > > 
> > > 
> > > go to venv\scripts open powershell\terminal and type activate and it will show (venv) before directory then try again, or type pip show torch-directml and post what version it shows you
> > 
> > 
> > Sorry, I am quite new to all this. What do you exactly mean by "powershell/terminal", I don't quite understand how I should "activate" scripts through them. And is it correct that after I "activated" scripts I need to run the webui-user.bat with the code:
> > ```
> > pip install torch-directml==0.1.13.1.dev230413
> > @echo off
> > 
> > set PYTHON=
> > set GIT=
> > set VENV_DIR=
> > set COMMANDLINE_ARGS=
> > 
> > call webui.bat
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Thanks in advance.
> 
> Don't put that in the webui-user.bat, if that's what you did. In the `stable-diffusion-webui-directml` there is a folder called `venv` and inside that, one called `scripts`, open the scripts folder, right click or shift + right click in the folder in some empty space, and it should have open in terminal or open powershell here, it depends on what Windows version you have, after that type `.\activate` and hit enter. It should show **(venv)** before the folder name now, you activated the `venv` don't worry about what that means. Now you can try `pip install torch-directml==0.1.13.1.dev230413` ![venvscreen](https://user-images.githubusercontent.com/25047164/235808516-6ce42d63-2333-4450-8dc0-5d782821c764.jpg)
> 
> Hope my way of explaining it wasn't too complicated. Just keep in mind this might change and you may need to update again when the code owner straightens it out.

I am not entirely sure what I am doing wrong. When I launch the `web-user.bat`, I get an error about torch not being able to install correctly. I received help on Discord and I was told to type `pip install torch-directml==0.1.13.1.dev230413` in cmd inside the stable-diffusion-webui-directml directory, but that did not work. I followed the steps in this thread. I went to my `venv` folder and activated it using powershell. After activating it, I pasted `pip install torch-directml==0.1.13.1.dev230413` inside powershell and saw the download commencing. When that finished, I launched the `web-user.bat` and saw that it was downloading more this time and skipped my torch installation. I assumed it was going to work this time until I got another error. I was reading the most recent comments in this thread and people said deleting the venv folder worked for them, but it put me back to where I started which was cmd not being able to download torch. I have included screenshots of my errors below. Hopefully someone can help. 

(What happens when I launch the .bat without the venv folder)
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/130991172/47373870-61fe-49a3-aa5d-78e69dd64dca)

(What happens when I launch the .bat after venv has been activated) 
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/130991172/8827a980-1711-4f44-abb1-f95441180b58)

**Comment by imfine3keyou** [comment]
I think you need to turn off the automatic updates for web-ui.
Right-click on webui-user.bat and select Edit to see the following:
@echo off

set PYTHON=
set GIT=
set VENV_DIR=
set COMMANDLINE_ARGS=--opt-sub-quad-attention --no-half --precision full --medvram

call webui.bat
Remove the automatic update under set COMMANDLINE_ARGS=-opt-sub-quad-attention --no-half --precision full --medvram.
becomes the following:
@echo off

set PYTHON=
set GIT=
set VENV_DIR=
set COMMANDLINE_ARGS=--opt-sub-quad-attention --no-half --precision full --medvram

call webui.bat
Then try again to delete the venv file and start webui-user.bat again
I don't know if you can understand it because I don't know code and English is not my native language.

**Comment by TheAkashicTraveller** [comment]
Okay so I was having the same issue and then I downgraded with `pip install  torch-directml==0.1.13.1.dev230413` and then ran the `pip install -r requirements.txt` and now it works. `Found existing installation: torch 2.0.0` hang on a second that just put me back on the same version of torch that I started on, but it works now? Also for me this started when I fetched the latest from this repo.

**Comment by tw-raZ** [comment]
This thread has been a blessing. I had this problem, the venv activate didn't work. I did the requirements.txt install and then deleted my venv folder and ran webui-user.bat. Worked! Thank you all contributors to this thread

**Comment by Natotela** [comment]
> 

```
F:\progz\stable-diffusion-webui-directml> .\venv\Scripts\activate
(venv) PS F:\progz\stable-diffusion-webui-directml> pip install torch-directml==0.1.13.1.dev230413
ERROR: Could not find a version that satisfies the requirement torch-directml==0.1.13.1.dev230413 (from versions: none)
ERROR: No matching distribution found for torch-directml==0.1.13.1.dev230413
```

**Comment by Natotela** [comment]
> Same issue here. Had to activate my venv `.\venv\Scripts\activate` and manualy downgrade torch-directml with `pip install torch-directml==0.1.13.1.dev230413` to fix it.

```
F:\progz\stable-diffusion-webui-directml> .\venv\Scripts\activate
(venv) PS F:\progz\stable-diffusion-webui-directml> pip install torch-directml==0.1.13.1.dev230413
ERROR: Could not find a version that satisfies the requirement torch-directml==0.1.13.1.dev230413 (from versions: none)
ERROR: No matching distribution found for torch-directml==0.1.13.1.dev230413
```

after cloning/1st-run-installing, I added "--backend directml" and got the py error of wtf is directml

**Comment by Natotela** [comment]
> Which version of python did you use?

fair question but I am not sure at the moment, I guess it was Python 3.10.9, assuming I tried using the same venv as SD auto1111

**Comment by lshqqytiger** [comment]
`torch-directml` should be available on Windows AMD64 Python 3.10 (cp310).
I think removing venv folder can be simple solution.

**Comment by getsmokd** [comment]
I am having the same issue, where is tells me "could not find a version that satisfies the requirements"

I followed every tip in this list but none of them made any difference. I don't understand why I cant tell it to install the correct version of torch and be done with it. Or even better, why can it not just install the correct version if it can tell I don't have it installed?
